snap/validate.go

```diff
@@ -30,89 +30,32 @@ import (
 	"strings"
 	"unicode/utf8"
 
+	"github.com/snapcore/snapd/snap/naming"
 	"github.com/snapcore/snapd/spdx"
 	"github.com/snapcore/snapd/strutil"
 	"github.com/snapcore/snapd/timeutil"
 )
 
-// Regular expressions describing correct identifiers.
-var validHookName = regexp.MustCompile("^[a-z](?:-?[a-z0-9])*$")
-
 // The fixed length of valid snap IDs.
 const validSnapIDLength = 32
 
-// almostValidName is part of snap and socket name validation.
-//   the full regexp we could use, "^(?:[a-z0-9]+-?)*[a-z](?:-?[a-z0-9])*$", is
-//   O(2ⁿ) on the length of the string in python. An equivalent regexp that
-//   doesn't have the nested quantifiers that trip up Python's re would be
-//   "^(?:[a-z0-9]|(?<=[a-z0-9])-)*[a-z](?:[a-z0-9]|-(?=[a-z0-9]))*$", but Go's
-//   regexp package doesn't support look-aheads nor look-behinds, so in order to
-//   have a unified implementation in the Go and Python bits of the project
-//   we're doing it this way instead. Check the length (if applicable), check
-//   this regexp, then check the dashes.
-//   This still leaves sc_snap_name_validate (in cmd/snap-confine/snap.c) and
-//   snap_validate (cmd/snap-update-ns/bootstrap.c) with their own handcrafted
-//   validators.
-var almostValidName = regexp.MustCompile("^[a-z0-9-]*[a-z][a-z0-9-]*$")
-
-// validInstanceKey is a regular expression describing valid snap instance key
-var validInstanceKey = regexp.MustCompile("^[a-z0-9]{1,10}$")
-
-// isValidName checks snap and socket socket identifiers.
-func isValidName(name string) bool {
-	if !almostValidName.MatchString(name) {
-		return false
-	}
-	if name[0] == '-' || name[len(name)-1] == '-' || strings.Contains(name, "--") {
-		return false
-	}
-	return true
-}
-
 // ValidateInstanceName checks if a string can be used as a snap instance name.
 func ValidateInstanceName(instanceName string) error {
-	// NOTE: This function should be synchronized with the two other
-	// implementations: sc_instance_name_validate and validate_instance_name .
-	pos := strings.IndexByte(instanceName, '_')
-	if pos == -1 {
-		// just store name
-		return ValidateName(instanceName)
-	}
-
-	storeName := instanceName[:pos]
-	instanceKey := instanceName[pos+1:]
-	if err := ValidateName(storeName); err != nil {
-		return err
-	}
-	if !validInstanceKey.MatchString(instanceKey) {
-		return fmt.Errorf("invalid instance key: %q", instanceKey)
-	}
-	return nil
+	return naming.ValidateInstance(instanceName)
```

Comment:
what are you thoughts? is that a intermediate step for churn? I would suppose in the end we don't want this indirection functions

Reply:
Yes, I didn't want the diff to be overly large. The cleanup can be introduced once this PR lands.